url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/Triangle3D.java

```diff
@@ -0,0 +1,3211 @@
+	 *             - if index of a triangle point is not 0, 1 or 2. The
+	 *             exception originates in the method getPoint(int) of the class
+	 *             Triangle3D.
+	 * @throws ArithmeticException
+	 *             - if norm equals zero in epsilon range. This exception
+	 *             originates in the method normalize(ScalarOperator) of the
+	 *             class Vector3D.
+	 */
+	public SimpleGeoObj intersection(Segment3D segment, ScalarOperator sop) { // Dag
+
+		SimpleGeoObj result = this.intersection(segment.getLine(sop), sop);
+
+		if (result == null)
+			return null;
+
+		if (result.getType() == SimpleGeoObj.POINT3D) {
+			if (segment.containsOnLine(((Point3D) result), sop))
+				return result;
+			else
+				return null;
+		} else
+			// result is segment, thus intersectionInPlane
+			return this.intersectionInPlane(segment, sop);
+	}
+
+	/**
+	 * Computes the intersection of this and the given triangle.<br>
+	 * Returns <code>null</code> if no intersection occures.<br>
+	 * Returns a Point3D object if the triangles touch in a point.<br>
+	 * Returns a Segment3D if the triangles intersect in a segment.<br>
+	 * Returns a Triangle3D object if one triangle contains the other one or
+	 * they do intersect in one.<br>
+	 * Returns a PointSet3D object if the triangles intersect in more than three
+	 * intersection points (4 to 6 are possible).<br>
+	 * In this case a new triangulation step for the resulting PointSet has to
+	 * be done. and this results in a more komplex surface than triangle<br>
+	 * 
+	 * @param triangle
+	 *            Triangle3D for intersection
+	 * @param sop
+	 *            ScalarOperator
+	 * @return SimpleGeoObj - resulting object. May be <code>null</code>,
+	 *         <code>Point3D</code>, <code>Segment3D</code>,
+	 *         <code>Trangle3D</code> or <code>PointSet3D</code>.
+	 * @throws IllegalStateException
+	 *             - if the intersectsInt(Line3D line, ScalarOperator sop)
+	 *             method (which computes the intersection of two lines) of the
+	 *             class Line3D returns a value that is not -2, -1, 0 or 1.
+	 * @throws IllegalStateException
+	 *             - signals Problems with the dimension of the wireframe.
+	 * @throws IllegalArgumentException
+	 *             - if validation of a Triangle3D fails. The exception
+	 *             originates in the constructor Triangle3D(Point3D, Point3D,
+	 *             Point3D, ScalarOperator).
+	 * @throws IllegalArgumentException
+	 *             - if index of a triangle point is not 0, 1 or 2. The
+	 *             exception originates in the method getPoint(int) of the class
+	 *             Triangle3D.
+	 * @throws ArithmeticException
+	 *             - if norm equals zero in epsilon range. This exception
+	 *             originates in the method normalize(ScalarOperator) of the
+	 *             class Vector3D.
+	 */
+
+	public SimpleGeoObj intersection(Triangle3D triangle, ScalarOperator sop) { // Dag
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 2446-2498, lines 2500-2516 into separate methods.